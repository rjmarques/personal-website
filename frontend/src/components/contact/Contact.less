@import "../../theme.less";

.contact {
  .contact-form {
    overflow: hidden;
    display: grid;
    row-gap: 8px;
    row-gap: auto;
    grid-template-columns: 1fr;

    .for-size(laptop-small-up; {
      grid-template-columns: repeat(2, 1fr);
      column-gap: 32px;
      row-gap: 16px;
    });
  }

  .full-row {
    .for-size(laptop-small-up; {
      grid-column-start: 1;
      grid-column-end: 3;
    });
  }

  .contact-text-area {
    font-size: 16px;
  }

  .ant-form-item:not(.ant-form-item-has-error) {
    textarea:not(:focus),
    input:not(:focus) {
      border-color: @light-grey;
    }
  }
}
