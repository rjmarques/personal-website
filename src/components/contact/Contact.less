@import "../../theme.less";

.Contact {
  .Contact-form-column {
    margin-bottom: 12px;
  }

  .Contact-text-area {
    font-size: 16px;
  }

  .ant-form-item-control:not(.has-error) {
    textarea:not(:focus),
    input:not(:focus) {
      border-color: @light-grey;
    }
  }

  .Contact-submit {
    margin-top: 15px;
  }
}
