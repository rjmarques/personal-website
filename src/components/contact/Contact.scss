@use "../../theme.scss" as *;

.contact {
  .contact-form {
    overflow: hidden;
    
    .form-grid {
      display: grid;
      grid-template-columns: 1fr;
      row-gap: 12px;
      
      @include for-size(laptop-small-up) {
        grid-template-columns: repeat(2, 1fr);
        column-gap: 32px;
        row-gap: 16px;
      }
      
      .full-width {
        @include for-size(laptop-small-up) {
          grid-column-start: 1;
          grid-column-end: 3;
        }
      }
    }

    .contact-text-area {
      font-size: 16px;
    }

    .submit-button {
      font-weight: 600;
      margin-bottom: 24px
    }

    .captcha-container {
      padding-bottom: 18px;

      .captcha-error {
        color: #d32f2f;
        font-size: 0.75rem;
        margin-left: 14px;
        position: absolute;
      }
    }
  }  
}
