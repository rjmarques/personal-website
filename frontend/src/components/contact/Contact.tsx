import { Button, Form, Input, message } from "antd";
import { FormInstance } from "antd/lib/form";
import React, { Component } from "react";

import ContentSection from "../common/ContentSection";
import ReCaptchaInput from "./ReCaptchaInput";

import "./Contact.less";

const FormItem = Form.Item;

interface IProps {
  sendMessage: (
    name: string,
    email: string,
    message: string,
    captcha: string,
    subject?: string,
    company?: string
  ) => Promise<void>;
}

interface IState {
  isSending: boolean;
}

class Contact extends Component<IProps, IState> {
  private formRef = React.createRef<FormInstance>();

  constructor(props: IProps) {
    super(props);

    this.state = {
      isSending: false,
    };
  }

  public onFinish = async (values: any) => {
    const hide = message.loading("Sending message...", 0);
    this.setState({ isSending: true });

    console.log(values);

    try {
      await this.props.sendMessage(
        values.name,
        values.email,
        values.message,
        values.captcha,
        values.subject,
        values.company
      );

      this.formRef.current!.resetFields([
        "name",
        "email",
        "message",
        "subject",
        "company",
        "captcha",
      ]);

      message.success(
        "Thank you for your message! I'll come back to you as soon as possible.",
        5
      );
    } catch (error) {
      console.log(error);
      message.error("Uh oh...Please refresh the page and try again", 5);
    } finally {
      hide();
      this.setState({ isSending: false });
    }
  };

  public onFinishFailed = (errorInfo: any) => {
    console.log("Failed:", errorInfo);
  };

  public validateReCaptcha = (rule: any, value: any) => {
    return value
      ? Promise.resolve()
      : Promise.reject("You're not a bot, are you?");
  };

  public render() {
    return (
      <ContentSection
        title="Contact"
        subTitle="Want to get in touch? Leave me a message"
        wrapperClass="contact"
      >
        <Form
          ref={this.formRef}
          name="contact-form"
          onFinish={this.onFinish}
          onFinishFailed={this.onFinishFailed}
          className="contact-form materialize"
        >
          <FormItem
            name="name"
            rules={[
              {
                message: "What's your name?",
                required: true,
              },
            ]}
          >
            <Input size="large" placeholder="Your Name *" />
          </FormItem>
          <FormItem
            name="email"
            rules={[
              {
                message: "The input is not valid Email",
                type: "email",
              },
              {
                message: "How do I message you back?",
                required: true,
              },
            ]}
          >
            <Input size="large" type="text" placeholder="Email *" />
          </FormItem>
          <FormItem name="subject">
            <Input size="large" placeholder="Subject" />
          </FormItem>
          <FormItem name="company">
            <Input size="large" type="text" placeholder="Company" />
          </FormItem>
          <FormItem
            className="full-row"
            name="message"
            rules={[
              {
                message: "Did you want to tell me something?",
                required: true,
              },
            ]}
          >
            <Input.TextArea
              className="contact-text-area"
              placeholder="Message *"
              rows={6}
            />
          </FormItem>
          <FormItem
            name="captcha"
            rules={[{ validator: this.validateReCaptcha }]}
            initialValue={undefined}
          >
            <ReCaptchaInput />
          </FormItem>
          <FormItem className="full-row">
            <Button
              type="primary"
              htmlType="submit"
              size="large"
              loading={this.state.isSending}
            >
              send message
            </Button>
          </FormItem>
        </Form>
      </ContentSection>
    );
  }
}

export default Contact;
