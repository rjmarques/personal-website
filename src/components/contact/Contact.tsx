import React, { useState } from 'react';
import {
  Button,
  TextField,
  Snackbar,
  Alert
} from '@mui/material';

import ContentSection from '../common/ContentSection';
import TurnstileInput from './TurnstileInput';
import './Contact.scss';

interface ContactProps {
  sendMessage: (
    name: string,
    email: string,
    message: string,
    captcha: string,
    subject?: string,
    company?: string
  ) => Promise<void>;
}

interface FormData {
  name: string;
  email: string;
  subject: string;
  company: string;
  message: string;
  captcha: string;
}

interface FormErrors {
  name?: string;
  email?: string;
  subject?: string;
  company?: string;
  message?: string;
  captcha?: string;
}

const Contact: React.FC<ContactProps> = ({ sendMessage }) => {
  const [formData, setFormData] = useState<FormData>({
    name: '',
    email: '',
    subject: '',
    company: '',
    message: '',
    captcha: ''
  });

  const [errors, setErrors] = useState<FormErrors>({});
  const [isSending, setIsSending] = useState(false);
  const [snackbar, setSnackbar] = useState<{
    open: boolean;
    message: string;
    severity: 'success' | 'error';
  }>({
    open: false,
    message: '',
    severity: 'success'
  });

  const validateForm = (): boolean => {
    const newErrors: FormErrors = {};

    if (!formData.name.trim()) {
      newErrors.name = "What's your name?";
    }

    if (!formData.email.trim()) {
      newErrors.email = "How do I message you back?";
    } else if (!/\S+@\S+\.\S+/.test(formData.email)) {
      newErrors.email = "The input is not valid Email";
    }

    if (!formData.message.trim()) {
      newErrors.message = "Did you want to tell me something?";
    }

    if (!formData.captcha) {
      newErrors.captcha = "You're not a bot, are you?";
    }

    setErrors(newErrors);
    return Object.keys(newErrors).length === 0;
  };

  const handleInputChange = (field: keyof FormData) => (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    const value = event.target.value;
    setFormData(prev => ({ ...prev, [field]: value }));
    
    // Clear error when user starts typing
    if (errors[field]) {
      setErrors(prev => ({ ...prev, [field]: undefined }));
    }
  };

  const handleCaptchaChange = (token: string | null) => {
    console.log('Captcha token:', token);
    setFormData(prev => ({ ...prev, captcha: token || '' }));
    
    // Clear captcha error when user completes the challenge
    if (errors.captcha) {
      setErrors(prev => ({ ...prev, captcha: undefined }));
    }
  };



  const handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();
    
    if (!validateForm()) {
      return;
    }

    setIsSending(true);

    try {
      await sendMessage(
        formData.name,
        formData.email,
        formData.message,
        formData.captcha,
        formData.subject,
        formData.company
      );

      // Reset form
      setFormData({
        name: '',
        email: '',
        message: '',
        captcha: '',
        subject: '',
        company: ''
      });

      setSnackbar({
        open: true,
        message: "Thank you for your message! I'll come back to you as soon as possible.",
        severity: 'success'
      });
    } catch (error) {
      console.log(error);
      setSnackbar({
        open: true,
        message: "Uh oh...Please refresh the page and try again",
        severity: 'error'
      });
    } finally {
      setIsSending(false);
    }
  };

  const handleCloseSnackbar = () => {
    setSnackbar(prev => ({ ...prev, open: false }));
  };

  return (
    <ContentSection
      title="Contact"
      subTitle="Want to get in touch? Leave me a message"
      wrapperClass="contact"
    >
      <form onSubmit={handleSubmit} className="contact-form">
        <div className="form-grid">
          <TextField
            className="input-field"
            fullWidth
            size="small"
            placeholder="Your Name *"
            value={formData.name}
            onChange={handleInputChange('name')}
            error={!!errors.name}
            helperText={errors.name || ' '}
            disabled={isSending}
          />
          <TextField
            className="input-field"
            fullWidth
            size="small"
            type="email"
            placeholder="Email *"
            value={formData.email}
            onChange={handleInputChange('email')}
            error={!!errors.email}
            helperText={errors.email || ' '}
            disabled={isSending}
          />
          <TextField
            className="input-field"
            fullWidth
            size="small"
            placeholder="Subject"
            value={formData.subject}
            onChange={handleInputChange('subject')}
            helperText={' '}
            disabled={isSending}
          />
          <TextField
            className="input-field"
            fullWidth
            size="small"
            placeholder="Company"
            value={formData.company}
            onChange={handleInputChange('company')}
            helperText={' '}
            disabled={isSending}
          />
          <TextField
            className="input-field contact-text-area full-width"
            fullWidth
            multiline
            rows={6}
            placeholder="Message *"
            value={formData.message}
            onChange={handleInputChange('message')}
            error={!!errors.message}
            helperText={errors.message || ' '}
            disabled={isSending}
          />
          <div className="captcha-container full-width">
            <TurnstileInput
              onChange={handleCaptchaChange}
            />
            {errors.captcha && (
              <div className="captcha-error">
                {errors.captcha}
              </div>
            )}
          </div>
          <div className='full-width'>
            <Button
                type="submit"
                variant="contained"
                size="large"
                disabled={isSending}
                loading={isSending}
                className='submit-button'
            >
                {isSending ? 'Sending...' : 'send message'}
            </Button>
          </div>
        </div>
      </form>

      <Snackbar
        open={snackbar.open}
        autoHideDuration={5000}
        onClose={handleCloseSnackbar}
        anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
      >
        <Alert 
          onClose={handleCloseSnackbar} 
          severity={snackbar.severity}
          sx={{ width: '100%' }}
        >
          {snackbar.message}
        </Alert>
      </Snackbar>
    </ContentSection>
  );
};

export default Contact;
