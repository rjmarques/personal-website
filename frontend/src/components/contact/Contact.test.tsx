import React from "react";
import renderer from "react-test-renderer";

import Contact from "./Contact";

describe("renders without crashing", () => {
  it("renders correctly", async () => {
    const sendMessage = jest.fn();
    const component = renderer.create(<Contact sendMessage={sendMessage} />);
    expect(component).toMatchSnapshot();
  });
});
