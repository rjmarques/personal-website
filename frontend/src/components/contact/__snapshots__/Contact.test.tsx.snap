// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing renders correctly 1`] = `
<section
  className="contact"
>
  <div
    className="content"
  >
    <div
      className="content-title"
    >
      <h2>
        Contact
      </h2>
      <span
        className="section-sub-title"
      >
        Want to get in touch? Leave me a message
      </span>
    </div>
    <form
      className="ant-form ant-form-horizontal contact-form materialize"
      id="contact-form"
      onSubmit={[Function]}
    >
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <input
                className="ant-input ant-input-lg"
                id="contact-form_name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Your Name *"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <input
                className="ant-input ant-input-lg"
                id="contact-form_email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Email *"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <input
                className="ant-input ant-input-lg"
                id="contact-form_subject"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Subject"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <input
                className="ant-input ant-input-lg"
                id="contact-form_company"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                placeholder="Company"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item full-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <textarea
                className="ant-input contact-text-area"
                id="contact-form_message"
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="Message *"
                rows={6}
                style={Object {}}
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <div />
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-row ant-form-item full-row"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            >
              <button
                className="ant-btn ant-btn-primary ant-btn-lg"
                onClick={[Function]}
                type="submit"
              >
                <span>
                  send message
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</section>
`;
